### PR TITLE
NEW funclionality to Clone user with permissions and categories

### PR DESCRIPTION
**Pull request by lamrani002**
# NEW|New functionality to clone user

### CODE REVIEW

File: htdocs/user/class/user.class.php [hunk 1]
```diff
@@ -4203,4 +4202,87 @@ public function findUserIdByEmail($email)
...
+		// Construction de la requête d'insertion
+		$sql = "INSERT INTO ".$this->db->prefix()."user_rights (entity, fk_user, fk_id)";
+		$sql .= " SELECT entity, ".$this->db->escape($toId).", fk_id";
+		$sql .= " FROM ".$this->db->prefix()."user_rights src";
+		$sql .= " WHERE fk_user = ".((int) $fromId);
+		$sql .= " AND NOT EXISTS (";
+		$sql .= "   SELECT 1";
+		$sql .= "   FROM ".$this->db->prefix()."user_rights dest";
+		$sql .= "   WHERE dest.entity = src.entity";
+		$sql .= "   AND dest.fk_user = ".$this->db->escape($toId);
```
Comment (eldy) [inline htdocs/user/class/user.class.php:4240]: It is an int so must use
$sql .= " AND dest.fk_user = ".((int) $toId);

File: htdocs/user/class/user.class.php [hunk 2]
```diff
@@ -4203,4 +4202,87 @@ public function findUserIdByEmail($email)
...
+		dol_syslog(get_class($this)."::clone_rights (delete default permissions)", LOG_DEBUG);
+
+		if (!$this->db->query($sqlDelete)) {
+			$this->db->rollback();
+			return -1;
+		}
+
+		// Construction de la requête d'insertion
+		$sql = "INSERT INTO ".$this->db->prefix()."user_rights (entity, fk_user, fk_id)";
+		$sql .= " SELECT entity, ".$this->db->escape($toId).", fk_id";
```
Comment (eldy) [inline htdocs/user/class/user.class.php:4233]: It is an int so must use
$sql .= " SELECT entity, ".((int) $toId).", fk_id";

**Comment by hregis** [comment]
@lamrani002 it will probably be necessary to adapt the cloning functions when Multicompany uses user centralization on the main entity!

**Comment by eldy** [comment]
> @lamrani002 it will probably be necessary to adapt the cloning functions when Multicompany uses user centralization on the main entity!

An easy to do first step can be to show the clone feature only if user is in same entity than current entity. So cloning is kept always in same entity.

**Pull request merged**